### DISCUSSION TITLE
stdError.frm

### DISCUSSION THREAD

**Discussion by Almesi** [sancarn/stdVBA#107] [Ideas]
Hey there,

i know you already have a std.Error.cls in WIP.
Ive made my own, which is currently only a Proof of Concept.
As i am too dumb to send you this via a Request ill add the link to my own repository:
Doc:
https://github.com/Almesi/AL_StdLib/tree/main/Doc/Form

Src:
https://github.com/Almesi/AL_StdLib/tree/main/Src/Form

Ill continue working on it and i will add more features.
Are you interested in it?

**Comment by sancarn** [comment]
Heya, a console would be cool, but also I don't think this makes much sense as a core part of the library. I do think as one of [`stdVBA-examples`](https://github.com/sancarn/stdVBA-examples) this does make a lot of sense though, if it were built on top of stdVBA technology. I would make a few changes too:

1. I would evaluate using `stdLambda` - as such you could very simply write 1+1 to perform calculations
2. I would suggest setting it up such that functions need to be registered by default, and a wider "open" setting were functions can be used on-demand.
3. Ability to use coloured text would be awesome
4. I'd likely use `stdUIElement` to control form elements too.

So yes, in general would be interested but not part of the core library imo.

`stdError` for stdVBA though is meant to be a simple wrapper / linker for 1. building stack traces and 2. printing errors to peripherals (e.g. immediate window, html page, file etc.)

**Reply by Almesi** [reply]
Thank you for your input. I will look into it

**Reply by Almesi** [reply]
> 2. I would suggest setting it up such that functions need to be registered by default, and a wider "open" setting were functions can be used on-demand.
What exactly do you mean here? Intellisense?

**Reply by sancarn** [reply]
@Almesi See [BindGlobal](https://github.com/sancarn/stdVBA/blob/master/src/stdLambda.cls#L293-L309). If a global implements `stdICallable` interface, it will act as a method within lambda engine. I.E. it can be called! Thus allowing the below:

```vb
Type TThis
  expressionEngine as stdLambda
End Type

'e.g. Console.Create(Array( _
'  "MyMethod", stdCallback.CreateFromModule("MyModule","MyMethod"), _
'  "Save", stdCallback.CreateFromObjectMethod(ThisWorkbook, "Save") _
'))
Public Sub protInit(ByVal registeredFunctions())
  For i = LBound(registeredFunctions) to UBound(registeredFunctions) step 2
      this.expressionEngine.bindGlobal(registeredFunctions(i), registeredFunctions(i+1))
  next
End Sub
```

To then allow for running of expressions like:

```vb
MyMethod(1,2,"hello world")
Save()
```

This act of pre-registering is, I think, important as it's a security measure. `stdLambda` is sandboxed by default.
